File: door_handle_detector/test/point_cloud_cropper.cpp
/*
 *  Copyright (c) 2008, Willow Garage, Inc.
 *  All rights reserved.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

/* Author: Brian Gerkey */

#include <cstdio>

#include "ros/ros.h"
#include "ros/console.h"

#include "tf/transform_listener.h"

#include "sensor_msgs/PointCloud.h"
#include "sensor_msgs/ChannelFloat32.h"
#include "geometry_msgs/Point32.h"

class Cropper
{
  public:
    Cropper(double zmin, double zmax, const std::string& frame_id)
    {
      zmin_ = zmin;
      zmax_ = zmax;
      frame_id_ = frame_id;

      sub_ = n.subscribe<sensor_msgs::PointCloud>("full_cloud", 50, &Cropper::cloudCallback, this);
      pub_ = n.advertise<sensor_msgs::PointCloud>("full_cloud_cropped", 1);
    }

    ~Cropper()
    {
    }
  private:
    ros::NodeHandle n;
    tf::TransformListener tf_;
    ros::Subscriber sub_;
    ros::Publisher pub_;

    double zmin_, zmax_;
    std::string frame_id_;

    void cloudCallback(const sensor_msgs::PointCloudConstPtr& cloud_in)
    {
      ROS_INFO("Received message with %d points", (int)cloud_in->points.size());
      sensor_msgs::PointCloud cloud_transformed;


      if (!tf_.waitForTransform(frame_id_, cloud_in->header.frame_id, cloud_in->header.stamp, ros::Duration(2.0))) return
      tf_.transformPointCloud(frame_id_, *cloud_in, cloud_transformed);

      sensor_msgs::PointCloud cloud_out;
      cloud_out.header = cloud_transformed.header;
      for(unsigned int j=0;j<cloud_transformed.channels.size();j++)
      {
        sensor_msgs::ChannelFloat32 c;
        c.name = cloud_transformed.channels[j].name;
        cloud_out.channels.push_back(c);
      }
      for(unsigned int i=0;i<cloud_transformed.points.size();i++)
      {
        if((cloud_transformed.points[i].z >= zmin_) &&
           (cloud_transformed.points[i].z <= zmax_))
        {
          cloud_out.points.push_back(cloud_transformed.points[i]);
          for(unsigned int j=0;j<cloud_transformed.channels.size();j++)
          {
            cloud_out.channels[j].values.push_back(cloud_transformed.channels[j].values[i]);
          }
        }
      }

      pub_.publish(cloud_out);
      ROS_INFO("Published message with %d points", (int)cloud_out.points.size());
    }
};

#define USAGE "USAGE point_cloud_cropper <zmin> <zmax> <frame>"

int
main(int argc, char** argv)
{
  ros::init(argc, argv, "cropper", ros::init_options::AnonymousName);
  if (argc != 4)
  {
    puts(USAGE);
    return 1;
  }

  Cropper c(atof(argv[1]), atof(argv[2]), argv[3]);
  ros::spin();
  return 0;
}

File: door_handle_detector/scripts/point_cloud_cropper.py
#!/usr/bin/env python
# Software License Agreement (BSD License)
#
# Copyright (c) 2008, Willow Garage, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of the Willow Garage nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#
# Revision $Id: add_two_ints_client 3357 2009-01-13 07:13:05Z jfaustwg $

# Crop a point cloud to Z min/max, in a given frame.  This could be more
# widely useful, but it turns out out to be really slow in Python.

PKG = 'door_handle_detector' # this package name

import roslib; roslib.load_manifest(PKG) 

import sys
from optparse import OptionParser
import tf
import bullet

import rospy
from sensor_msgs.msg import PointCloud, ChannelFloat32, Point32
from geometry_msgs.msg import Point32

NAME='point_cloud_cropper'

class Cropper:
  def __init__(self, zmin, zmax, frame):
    self.zmin = zmin
    self.zmax = zmax
    self.frame = frame
    self.sub = rospy.Subscriber('full_cloud', PointCloud, self.cloud_cb)
    self.pub = rospy.Publisher('full_cloud_cropped', PointCloud)
    self.tf = tf.listener.TransformListener()
    rospy.init_node('cropper', anonymous=True)

  def cloud_cb(self, msg):
    print 'Received message with %d points'%(len(msg.points))
    newmsg = PointCloud(msg.header,[],[])
    for c in msg.chan:
      newmsg.channels.append(ChannelFloat32(c.name,[]))
    for i in range(0,len(msg.points)):
      if self.frame is None:
        p = msg.points[i]
      else:
        tfp = tf.PointStamped(bullet.Vector3(msg.points[i].x, msg.points[i].y, msg.points[i].z), 0, 0, msg.header.frame_id)
        tfpprime = self.tf.transform_point(self.frame, tfp)
        p = Point32(tfpprime.point.x(), tfpprime.point.y(), tfpprime.point.z())
      if p.z >= self.zmin and p.z <= self.zmax:
        newmsg.points.append(p)
        for j in range(0,len(msg.chan)):
          newmsg.channels[j].values.append(msg.channels[j].values[i])
    print 'Publishing message with %d points'%(len(newmsg.points))
    self.pub.publish(newmsg)

if __name__ == "__main__":

  parser = OptionParser(usage="usage: %prog [options]", prog=NAME)
  parser.add_option("", "--zmin",
                    dest="zmin", default=None,
                    help="Minimum z value to allow through")
  parser.add_option("", "--zmax",
                    dest="zmax", default=None,
                    help="Maximum z value to allow through")
  parser.add_option("-f", "--frame",
                    dest="frame", default=None,
                    help="Frame to transform to before cropping")

  (options, args) = parser.parse_args(sys.argv)
  if options.zmin is None or options.zmax is None:
    parser.error("must specify both zmin and zmax")
    parser.print_help()

  c = Cropper(float(options.zmin), float(options.zmax), options.frame)
  rospy.spin()
